oil_lang/expr_eval.py

```diff
@@ -1236,19 +1338,143 @@ def _EvalFuncCall(self, node):
+                raise error.Expr('index out of range', loc.Missing)
+
+          else:
+            raise error.InvalidType('expected Slice or Int', loc.Missing)
+
+      elif case(value_e.Tuple):
+        obj = cast(value.Tuple, UP_obj)
+        with tagswitch(index) as case2:
+          if case2(value_e.Slice):
+              index = cast(value.Slice, index)
+              step = 1
+              if index.step:
+                step = index.step.i
+
+              try:
+                  if index.lower and index.upper:
+                    return value.Tuple(obj.items[index.lower.i:index.upper.i:step])
+
+                  elif index.lower:
+                    return value.Tuple(obj.items[index.lower.i::step])
+
+                  elif index.upper:
+                    return value.Tuple(obj.items[:index.upper.i:step])
+
+                  else:
+                    # l[:] == l
+                    return obj
+
+              except IndexError:
+                # TODO: expr.Subscript has no error location
+                raise error.Expr('index out of range', loc.Missing)
+
+          elif case2(value_e.Int):
+              index = cast(value.Int, index)
+              try:
+                return obj.items[index.i]
+              except IndexError:
+                # TODO: expr.Subscript has no error location
+                raise error.Expr('index out of range', loc.Missing)
+
+          else:
+            raise error.InvalidType('expected Slice or Int', loc.Missing)
+
+      elif case(value_e.Str):
+        obj = cast(value.Str, UP_obj)
+        with tagswitch(index) as case2:
+          if case2(value_e.Slice):
+              index = cast(value.Slice, index)
+              step = 1
+              if index.step:
+                  step = 1
+
+              try:
+                  if index.lower and index.upper:
+                    return value.Str(obj.s[index.lower.i:index.upper.i:step])
+
+                  elif index.lower:
+                    return value.Str(obj.s[index.lower.i::step])
+
+                  elif index.upper:
+                    return value.Str(obj.s[:index.upper.i:step])
+
+                  else:
+                    # l[:] == l
+                    return obj
+
+              except IndexError:
+                # TODO: expr.Subscript has no error location
+                raise error.Expr('index out of range', loc.Missing)
+
+          elif case2(value_e.Int):
+              index = cast(value.Int, index)
+              try:
+                return value.Str(obj.s[index.i])
+              except IndexError:
+                # TODO: expr.Subscript has no error location
+                raise error.Expr('index out of range', loc.Missing)
+
+          else:
+            raise error.InvalidType('expected Slice or Int', loc.Missing)
+
+      elif case(value_e.Dict):
+        obj = cast(value.Dict, UP_obj)
+        if index.tag() != value_e.Str:
+            raise error.InvalidType('expected String', loc.Missing)
```

Comment:
How about 'Expected string index for Dict' to be more precise

And it does remind me we need to add tokens to all the YSH AST nodes, similar to what Aidan and I have been doing for OSH